components/private_channel/rust/ffi/src/lib.rs

```diff
@@ -0,0 +1,257 @@
+            encoded_proofs_dec_size: 0,
+            random_vec: mock_vec.as_ptr(),
+            random_vec_size: 0,
+            encoded_proofs_rand: mock_vec.as_ptr(),
+            encoded_proofs_rand_size: 0,
+        }
+    }
+}
+
+/// Starts client attestation challenge;
+#[no_mangle]
+pub unsafe extern "C" fn client_start_challenge(
+    input: *const *const c_char,
+    input_size: c_int,
+    server_pk_encoded: *const c_char,
+) -> ResultChallenge {
+    abort_if_null!(input);
+    abort_if_null!(server_pk_encoded);
+
+    let server_pk = match CStr::from_ptr(server_pk_encoded).to_str() {
+        Ok(pk) => pk.to_string(),
+        Err(_) => return ResultChallenge::default(),
+    };
+
+    let mut v_out = Vec::new();
+    let input_array = slice::from_raw_parts(input, input_size as usize);
+    for n in 0..input_size {
+        let s_ptr = CStr::from_ptr(input_array[n as usize]);
+        v_out.push(s_ptr.to_str().unwrap().to_string());
+    }
+
+    let brave_private_channel::FirstRoundOutput {
+        pkeys,
+        skeys,
+        shared_pks,
+        enc_hashes,
+    } = match brave_private_channel::start_challenge(v_out, server_pk) {
+        Ok(result) => result,
+        Err(_) => return ResultChallenge::default(),
+    };
+
+    let pkeys_buff = pkeys.into_boxed_slice();
+    let pkeys_buff = std::mem::ManuallyDrop::new(pkeys_buff);
+
+    let skeys_buff = skeys.into_boxed_slice();
+    let skeys_buff = std::mem::ManuallyDrop::new(skeys_buff);
+
+    let shared_pks_buff = shared_pks.into_boxed_slice();
+    let shared_pks_buff = std::mem::ManuallyDrop::new(shared_pks_buff);
+
+    let enc_hashes_buff = enc_hashes.into_boxed_slice();
+    let enc_hashes_buff = std::mem::ManuallyDrop::new(enc_hashes_buff);
+
+    ResultChallenge {
+        pkeys: pkeys_buff.as_ptr(),
+        pkeys_byte_size: pkeys_buff.len(),
+        skeys: skeys_buff.as_ptr(),
+        skeys_byte_size: skeys_buff.len(),
+        shared_pubkey: shared_pks_buff.as_ptr(),
+        shared_pkeys_byte_size: shared_pks_buff.len(),
+        encrypted_hashes: enc_hashes_buff.as_ptr(),
+        encrypted_hashes_size: enc_hashes_buff.len(),
+        key_size: KEY_SIZE,
+        error: false,
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn client_second_round(
+    input: *const c_char,
+    _input_size: c_int,
+    client_sk_encoded: *const c_char,
+) -> ResultSecondRound {
+    abort_if_null!(input);
+    abort_if_null!(client_sk_encoded);
+
+    let client_sk = match CStr::from_ptr(client_sk_encoded).to_str() {
+        Ok(sk) => sk.to_string(),
+        Err(_) => return ResultSecondRound::default(),
+    };
+
+    let raw_str = match CStr::from_ptr(input).to_str() {
+        Ok(s) => s.to_string(),
+        Err(_) => return ResultSecondRound::default(),
+    };
+    let parsed_str = raw_str.replace(&['[', ']'][..], "");
+    let v_enc: Vec<u8> = parsed_str
+        .split(", ")
+        .map(|s| s.parse::<u8>().unwrap())
```

Comment:
Probably best to return `ResultSecondRound::default()` rather than panic on malformed data here.

`Result` implements `FromIterator`, which means you can actually collect the iterated `Result<u8, _>`s into a `Result<Vec<u8>, _>` and it will stop as soon as the first parse error is encountered.